src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs

```diff
@@ -103,6 +106,102 @@ internal sealed partial class LocalRewriter
+
+                Debug.Assert(right is not BoundConversion { Operand.Kind: BoundKind.TupleLiteral });
+                if (// was the RHS a tuple literal?
+                    right is { Kind: BoundKind.ConvertedTupleLiteral } or BoundConversion { Operand.Kind: BoundKind.ConvertedTupleLiteral }
+
+                    // were any of the RHS expressions stored into temps?
+                    && effects.init.Any()
+
+                    && canReorderTargetAssignments(lhsTargets, ref visitedSymbols))
+                {
+                    // Consider a deconstruction assignment like the following
+                    // (a, b, c) = (x, y, z);
+
+                    // (x, y, z) are evaluated into temps, then the temps are stored to the targets:
+                    // temp1 = x;
+                    // temp2 = y;
+                    // temp3 = z;
+                    // a = temp1;
+                    // b = temp2;
+                    // c = temp3;
+
+                    // In this code path, we have found a case where a, b, and c are non-side-effecting expressions which refer to unique/non-aliased variables.
+                    // In addition, at least one of the expressions x, y, or z actually needed to be evaluated into temps (versus being a constant value where the temp was elided at an earlier phase).
+                    // As an optimization, ensure that assignments from temps to targets happen in the reverse order of effects:
+                    // temp1 = x;
+                    // temp2 = y;
+                    // temp3 = z;
+                    // c = temp3;
+                    // b = temp2;
+                    // a = temp1;
+
+                    // This makes it more likely that the stack optimizer pass will be able to eliminate the temps and replace them with stack push/pops.
+                    // Note also that no optimization can occur if the expression being converted is not a literal, e.g. `(a, b, c) = M()`. Order of access won't affect our ability to eliminate temps in that case.
+                    effects.assignments.ReverseContents();
+                }
+
+                visitedSymbols.Dispose();
+            }
+
+            static bool canReorderTargetAssignments(ArrayBuilder<Binder.DeconstructionVariable> targets, ref TemporaryArray<Symbol> visitedSymbols)
+            {
+                // If we know all targets refer to distinct variables, then we can reorder the assignments.
+                // We avoid doing this in any cases where aliasing could occur, e.g.:
+                // var y = 1;
+                // ref var x = ref y;
+                // (x, y) = (a, b);
+
+                foreach (var target in targets)
+                {
+                    Debug.Assert(target is { Single: not null, NestedVariables: null } or { Single: null, NestedVariables: not null });
+                    if (target.Single is { } single)
+                    {
+                        Symbol? symbol = single switch
```

Comment:
>Symbol? symbol = single switch [](http://example.com/codeflow?start=24&length=30)

I understand that there is a desire to use new language features, but, in my opinion, a switch statement can incorporate logic from this statement and from the ```if``` below in a much clearer way. The logic won't be spread out across multiple conditional constructs, and it might actually be not much longer in terms of lines used. Up to you.